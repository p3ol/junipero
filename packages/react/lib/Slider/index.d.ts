import { ReactNode, ComponentPropsWithRef, MutableRefObject } from 'react';

export declare type SliderRef = {
  isJunipero: boolean;
  moving: boolean;
  precision: number;
  value: number;
  reset: () => void;
  fillRef: MutableRefObject<any>;
  handleRef: MutableRefObject<any>;
  innerRef: MutableRefObject<any>;
  slideRef: MutableRefObject<any>;
  tooltipRef: MutableRefObject<any>;
};

declare interface SliderProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  className?: string;
  disabled: boolean;
  max?: number;
  min?: number;
  maxValue?: number;
  minValue?: number;
  step?: number;
  tooltipEnabled?: boolean;
  value?: number;
  animateTooltip?(
    tooltip: ReactNode | JSX.Element,
    opts: { opened: boolean }
  ): ReactNode | JSX.Element;
  onMove?(value: number): void;
  parseTitle?(value: number): ReactNode | JSX.Element;
  ref?: MutableRefObject<SliderRef | undefined>;
}

declare function Slider(props: SliderProps): ReactNode | JSX.Element;

export default Slider;
