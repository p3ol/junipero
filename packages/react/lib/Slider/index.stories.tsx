import { action } from 'storybook/actions';
import { slideInDownMenu } from '@junipero/transitions';

import Slider from '.';

export default { title: 'react/Slider' };

export const basic = () => (
  <Slider onMove={action('move')} />
);

export const withoutTooltip = () => (
  <Slider tooltipEnabled={false} />
);

export const animatedTooltip = () => (
  <Slider animateTooltip={slideInDownMenu} />
);
