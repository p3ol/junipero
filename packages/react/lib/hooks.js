import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
