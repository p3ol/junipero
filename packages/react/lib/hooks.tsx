import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
