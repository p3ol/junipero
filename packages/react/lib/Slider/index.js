import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import {
  classNames,
  mockState,
  getFloatPrecision,
  ensureMinMax,
} from '@junipero/core';
import { useEventListener } from '@junipero/hooks';
import PropTypes from 'prop-types';

import Tooltip from '../Tooltip';

const Slider = forwardRef(({
  className,
  value = 0,
  disabled = false,
  globalEventsTarget = globalThis,
  min = 0,
  minValue = 0,
  max = 100,
  maxValue = 100,
  step = 1,
  tooltipEnabled = true,
  animateTooltip,
  onMove,
  onMouseDown,
  parseTitle = v => v,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const fillRef = useRef();
  const handleRef = useRef();
  const slideRef = useRef();
  const tooltipRef = useRef();
  const [state, dispatch] = useReducer(mockState, {
    value: parseFloat(
      ensureMinMax(Math.round(value / step) * step, minValue, maxValue
      ).toFixed(getFloatPrecision(step))),
    precision: getFloatPrecision(step),
    moving: false,
  });

  useEffect(() => {
    state.precision = getFloatPrecision(step);
    state.value = parseFloat(
      ensureMinMax(Math.round(value / step) * step, min, max)
        .toFixed(state.precision)
    );
    dispatch({ precision: state.precision, value: state.value });
  }, [value, step]);

  useImperativeHandle(ref, () => ({
    innerRef,
    fillRef,
    slideRef,
    handleRef,
    tooltipRef,
    value: state.value,
    moving: state.moving,
    precision: state.precision,
    reset,
    isJunipero: true,
  }));

  useEventListener('mousemove', e => {
    onMouseMove_(e);
  }, globalEventsTarget);

  useEventListener('mouseup', e => {
    onMouseUp_(e);
  }, globalEventsTarget);

  const onMouseDown_ = e => {
    if (e.button !== 0 || disabled) {
      return;
    }

    e.preventDefault();

    handleRef.current?.focus();
    tooltipRef.current?.open();
    state.moving = true;
    dispatch({ moving: true });
    onMouseMove_(e);
    onMouseDown?.(e);
  };

  const onMouseMove_ = e => {
    if (!state.moving || !slideRef.current || disabled) {
      return;
    }

    const previousValue = state.value;
    const offset = slideRef.current?.getBoundingClientRect();
    const width = slideRef.current?.offsetWidth;
    const position = max * (
      (e.pageX - (offset.left - document.body.scrollLeft)) / width
    ) || 0;

    state.value = parseFloat(
      Math.max(minValue, Math.min(maxValue, Math.round(position / step) * step))
        .toFixed(state.precision)
    );

    dispatch({ value: state.value });

    if (state.value !== previousValue) {
      onMove?.(state.value);
    }
  };

  const onMouseUp_ = () => {
    if (!state.moving) {
      return;
    }

    tooltipRef.current?.close();
    dispatch({ moving: false });
  };

  const onKeyDown_ = e => {
    if (e.key === 'ArrowLeft') {
      state.value = ensureMinMax(state.value - step, minValue, maxValue);
      dispatch({ value: state.value });
    } else if (e.key === 'ArrowRight') {
      state.value = ensureMinMax(state.value + step, minValue, maxValue);
      dispatch({ value: state.value });
    } else if (e.key === 'ArrowUp') {
      dispatch({ value: min });
    } else if (e.key === 'ArrowDown') {
      dispatch({ value: max });
    }
  };

  const reset = () => {
    dispatch({
      value: ensureMinMax(value, minValue, maxValue),
      precision: getFloatPrecision(step),
      moving: false,
    });
  };

  const getFillSize = () =>
    state.value / max;

  const getHandlePosition = () =>
    slideRef.current?.offsetWidth * getFillSize();

  return (
    <div
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'slider',
        { disabled },
        className,
      )}
      onMouseDown={onMouseDown_}
    >
      <div
        className="slide"
        ref={slideRef}
      >
        <div
          className="fill"
          ref={fillRef}
          style={{ width: `${getFillSize() * 100}%` }}
        />
      </div>
      <Tooltip
        ref={tooltipRef}
        text={parseTitle(state.value)}
        trigger="manual"
        disabled={!tooltipEnabled || disabled}
        animate={animateTooltip}
      >
        <div
          className={classNames(
            'handle',
            { moving: state.moving }
          )}
          ref={handleRef}
          tabIndex={0}
          style={{
            transform: `translate3d(${getHandlePosition()}px, 0, 0)`,
          }}
          onKeyDown={onKeyDown_}
        />
      </Tooltip>
    </div>
  );
});

Slider.displayName = 'Slider';
Slider.propTypes = {
  disabled: PropTypes.bool,
  globalEventsTarget: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.object,
  ]),
  max: PropTypes.number,
  maxValue: PropTypes.number,
  min: PropTypes.number,
  minValue: PropTypes.number,
  step: PropTypes.number,
  value: PropTypes.number,
  tooltipEnabled: PropTypes.bool,
  animateTooltip: PropTypes.func,
  onMove: PropTypes.func,
  onMouseDown: PropTypes.func,
  parseTitle: PropTypes.func,
};

export default Slider;
