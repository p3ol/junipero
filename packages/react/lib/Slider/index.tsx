import {
  ComponentPropsWithRef,
  MutableRefObject,
  ReactNode,
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import {
  classNames,
  mockState,
  getFloatPrecision,
  ensureMinMax,
} from '@junipero/core';
import { useEventListener } from '@junipero/hooks';
import PropTypes from 'prop-types';

import Tooltip from '../Tooltip';
import { ForwardedProps } from '../utils';

export declare type SliderRef = {
  isJunipero: boolean;
  moving: boolean;
  precision: number;
  value: number;
  reset: () => void;
  fillRef: MutableRefObject<any>;
  handleRef: MutableRefObject<any>;
  innerRef: MutableRefObject<any>;
  slideRef: MutableRefObject<any>;
  tooltipRef: MutableRefObject<any>;
};

export declare interface SliderProps extends ComponentPropsWithRef<any> {
  children?: ReactNode | JSX.Element;
  className?: string;
  disabled?: boolean;
  max?: number;
  min?: number;
  maxValue?: number;
  minValue?: number;
  step?: number;
  tooltipEnabled?: boolean;
  value?: number;
  animateTooltip?(
    tooltip: ReactNode | JSX.Element,
    opts: { opened: boolean }
  ): ReactNode | JSX.Element;
  onMove?(value: number): void;
  parseTitle?(value: number): ReactNode | JSX.Element;
  ref?: MutableRefObject<SliderRef | undefined>;
}

const Slider = forwardRef(({
  className,
  value = 0,
  disabled = false,
  globalEventsTarget = globalThis,
  min = 0,
  minValue = 0,
  max = 100,
  maxValue = 100,
  step = 1,
  tooltipEnabled = true,
  animateTooltip,
  onMove,
  onMouseDown,
  parseTitle = v => v,
  ...rest
}: SliderProps, ref) => {
  const innerRef = useRef<any>();
  const fillRef = useRef<any>();
  const handleRef = useRef<any>();
  const slideRef = useRef<any>();
  const tooltipRef = useRef<any>();
  const [state, dispatch] = useReducer(mockState, {
    value: parseFloat(
      ensureMinMax(Math.round(value / step) * step, minValue, maxValue
      ).toFixed(getFloatPrecision(step))),
    precision: getFloatPrecision(step),
    moving: false,
  });

  useEffect(() => {
    state.precision = getFloatPrecision(step);
    state.value = parseFloat(
      ensureMinMax(Math.round(value / step) * step, min, max)
        .toFixed(state.precision)
    );
    dispatch({ precision: state.precision, value: state.value });
  }, [value, step]);

  useImperativeHandle(ref, () => ({
    innerRef,
    fillRef,
    slideRef,
    handleRef,
    tooltipRef,
    value: state.value,
    moving: state.moving,
    precision: state.precision,
    reset,
    isJunipero: true,
  }));

  useEventListener('mousemove', e => {
    onMouseMove_(e);
  }, globalEventsTarget);

  useEventListener('mouseup', e => {
    onMouseUp_();
  }, globalEventsTarget);

  const onMouseDown_ = e => {
    if (e.button !== 0 || disabled) {
      return;
    }

    e.preventDefault();

    handleRef.current?.focus();
    tooltipRef.current?.open();
    state.moving = true;
    dispatch({ moving: true });
    onMouseMove_(e);
    onMouseDown?.(e);
  };

  const onMouseMove_ = e => {
    if (!state.moving || !slideRef.current || disabled) {
      return;
    }

    const previousValue = state.value;
    const offset = slideRef.current?.getBoundingClientRect();
    const width = slideRef.current?.offsetWidth;
    const position = max * (
      (e.pageX - (offset.left - document.body.scrollLeft)) / width
    ) || 0;

    state.value = parseFloat(
      Math.max(minValue, Math.min(maxValue, Math.round(position / step) * step))
        .toFixed(state.precision)
    );

    dispatch({ value: state.value });

    if (state.value !== previousValue) {
      onMove?.(state.value);
    }
  };

  const onMouseUp_ = () => {
    if (!state.moving) {
      return;
    }

    tooltipRef.current?.close();
    dispatch({ moving: false });
  };

  const onKeyDown_ = e => {
    if (e.key === 'ArrowLeft') {
      state.value = ensureMinMax(state.value - step, minValue, maxValue);
      dispatch({ value: state.value });
    } else if (e.key === 'ArrowRight') {
      state.value = ensureMinMax(state.value + step, minValue, maxValue);
      dispatch({ value: state.value });
    } else if (e.key === 'ArrowUp') {
      dispatch({ value: min });
    } else if (e.key === 'ArrowDown') {
      dispatch({ value: max });
    }
  };

  const reset = () => {
    dispatch({
      value: ensureMinMax(value, minValue, maxValue),
      precision: getFloatPrecision(step),
      moving: false,
    });
  };

  const getFillSize = () =>
    state.value / max;

  const getHandlePosition = () =>
    slideRef.current?.offsetWidth * getFillSize();

  return (
    <div
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'slider',
        { disabled },
        className,
      )}
      onMouseDown={onMouseDown_}
    >
      <div
        className="slide"
        ref={slideRef}
      >
        <div
          className="fill"
          ref={fillRef}
          style={{ width: `${getFillSize() * 100}%` }}
        />
      </div>
      <Tooltip
        ref={tooltipRef}
        text={parseTitle(state.value)}
        trigger="manual"
        disabled={!tooltipEnabled || disabled}
        animate={animateTooltip}
      >
        <div
          className={classNames(
            'handle',
            { moving: state.moving }
          )}
          ref={handleRef}
          tabIndex={0}
          style={{
            transform: `translate3d(${getHandlePosition()}px, 0, 0)`,
          }}
          onKeyDown={onKeyDown_}
        />
      </Tooltip>
    </div>
  );
}) as ForwardedProps<SliderProps, SliderRef>;

Slider.displayName = 'Slider';
Slider.propTypes = {
  disabled: PropTypes.bool,
  globalEventsTarget: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.object,
  ]),
  max: PropTypes.number,
  maxValue: PropTypes.number,
  min: PropTypes.number,
  minValue: PropTypes.number,
  step: PropTypes.number,
  value: PropTypes.number,
  tooltipEnabled: PropTypes.bool,
  animateTooltip: PropTypes.func,
  onMove: PropTypes.func,
  onMouseDown: PropTypes.func,
  parseTitle: PropTypes.func,
};

export default Slider;
