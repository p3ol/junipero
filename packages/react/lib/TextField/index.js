import {
  forwardRef,
  useEffect,
  useImperativeHandle,
  useReducer,
  useRef,
} from 'react';
import { classNames, mockState, exists } from '@junipero/core';
import PropTypes from 'prop-types';

import { useFieldControl } from '../hooks';

const TextField = forwardRef(({
  min,
  max,
  autoFocus,
  children,
  className,
  valid,
  value,
  disabled = false,
  required = false,
  type = 'text',
  onBlur,
  onChange,
  onFocus,
  onWheel,
  onValidate = (val, { required }) => !!val || !required,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const inputRef = useRef();
  const { update: updateControl } = useFieldControl();
  const [state, dispatch] = useReducer(mockState, {
    value: value ?? '',
    valid: valid ?? false,
    focused: autoFocus ?? false,
    dirty: false,
  });

  useImperativeHandle(ref, () => ({
    innerRef,
    inputRef,
    value: state.value,
    valid: state.valid,
    dirty: state.dirty,
    focused: state.focused,
    focus,
    blur,
    reset,
    setDirty,
    isJunipero: true,
  }));

  useEffect(() => {
    if (exists(value)) {
      state.value = value;
      state.valid = onValidate(state.value, { dirty: state.dirty, required });
      dispatch({ value: state.value, valid: state.valid });
      updateControl?.({ valid: state.valid, dirty: state.dirty });
    }
  }, [value]);

  useEffect(() => {
    dispatch({ valid: valid ?? false });
    updateControl?.({ valid: state.valid });
  }, [valid]);

  const parseValue = value => {
    const value_ = value.replace(',', '.');
    const parsedValue = parseFloat(value_);

    return !isNaN(parsedValue) ? parsedValue : value_;
  };

  const onChange_ = e => {
    /* html input disabled attribute will prevent onChange if present anyway */
    /* istanbul ignore next: just in case */
    if (disabled) {
      return;
    }

    if (type === 'number') {
      value = parseValue(e?.target?.value);
    } else {
      value = e?.target?.value;
    }

    state.value = value;
    state.dirty = true;
    state.valid = onValidate(
      state.value, { dirty: state.dirty, required }
    );
    dispatch({ value: state.value, dirty: state.dirty });
    onChange?.({ value: state.value, valid: state.valid, dirty: state.dirty });
    updateControl?.({ dirty: true, valid: state.valid });
  };

  const onFocus_ = e => {
    dispatch({ focused: true });
    updateControl?.({ focused: true });
    onFocus?.(e);
  };

  const onBlur_ = e => {
    dispatch({ focused: false });
    updateControl?.({ focused: false });
    onBlur?.(e);
  };

  const focus = () => {
    inputRef.current?.focus();
  };

  const blur = () => {
    inputRef.current?.blur();
  };

  const reset = () => {
    dispatch({ value: value ?? '', valid: valid ?? false, dirty: false });
    updateControl?.({ dirty: false, valid: valid ?? false });
  };

  const isEmpty = () =>
    !exists(state.value) || state.value === '';

  const setDirty = dirty => {
    state.dirty = dirty;
    dispatch({ dirty });
    updateControl?.({ dirty });
  };

  const Tag = rest.rows > 1 ? 'textarea' : 'input';

  return (
    <div
      ref={innerRef}
      className={classNames(
        'junipero',
        'text-field',
        state.dirty ? 'dirty' : 'pristine',
        !state.valid && state.dirty ? 'invalid' : 'valid',
        { empty: isEmpty(), focused: state.focused, disabled },
        className
      )}
    >
      <Tag
        { ...rest }
        min={min}
        max={max}
        autoFocus={autoFocus}
        className="field"
        ref={inputRef}
        disabled={disabled}
        required={required}
        type={Tag === 'textarea' ? undefined : type}
        value={state.value}
        onChange={onChange_}
        onFocus={onFocus_}
        onBlur={onBlur_}
        onWheel={onWheel ?? (() => inputRef.current?.blur())}
      />
      { children }
    </div>
  );
});

TextField.displayName = 'TextField';
TextField.propTypes = {
  min: PropTypes.number,
  max: PropTypes.number,
  autoFocus: PropTypes.bool,
  disabled: PropTypes.bool,
  required: PropTypes.bool,
  type: PropTypes.string,
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onWheel: PropTypes.func,
  onValidate: PropTypes.func,
};

export default TextField;
