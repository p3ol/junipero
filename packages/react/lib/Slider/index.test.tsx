import { createRef } from 'react';
import { render, fireEvent, act, createEvent } from '@testing-library/react';
import { vi } from 'vitest';

import Slider, { type SliderRef } from './';

describe('<Slider />', () => {
  it('should render', () => {
    const { container, unmount } = render(<Slider />);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should update precision when changing step', () => {
    const ref = createRef<SliderRef>();
    const { rerender, unmount } = render(<Slider ref={ref} />);
    expect(ref.current.precision).toBe(0);
    rerender(<Slider ref={ref} step={0.1} />);
    expect(ref.current.precision).toBe(1);
    unmount();
  });

  it('should return a value rounded to the correct step & precision', () => {
    const ref = createRef<SliderRef>();
    const { rerender, unmount } = render(
      <Slider ref={ref} min={0} max={100} step={1} />
    );
    rerender(
      <Slider ref={ref} min={0} max={100} step={1} value={15.6} />
    );
    expect(ref.current.value).toBe(16);

    rerender(
      <Slider ref={ref} min={0} max={100} step={1} value={-5} />
    );
    expect(ref.current.value).toBe(0);

    rerender(
      <Slider ref={ref} min={0} max={100} step={1} value={255.155435} />
    );
    expect(ref.current.value).toBe(100);

    rerender(
      <Slider ref={ref} min={0} max={100} step={0.1} value={15.6} />
    );
    expect(ref.current.value).toBe(15.6);

    rerender(
      <Slider ref={ref} min={0} max={100} step={0.1} value={15.68} />
    );
    expect(ref.current.value).toBe(15.7);

    rerender(
      <Slider ref={ref} min={0} max={100} step={0.1} value={-5.546} />
    );
    expect(ref.current.value).toBe(0);

    rerender(
      <Slider ref={ref} min={0} max={100} step={0.1} value={255.155435} />
    );
    expect(ref.current.value).toBe(100);

    rerender(
      <Slider ref={ref} min={0} max={1000} step={10} value={156} />
    );
    expect(ref.current.value).toBe(160);

    rerender(
      <Slider ref={ref} min={0} max={1000} step={10} value={-5.546} />
    );
    expect(ref.current.value).toBe(0);

    rerender(
      <Slider ref={ref} min={0} max={1000} step={10} value={255.155435} />
    );
    expect(ref.current.value).toBe(260);

    unmount();
  });

  it('should set slider as moving on mouse down', () => {
    const ref = createRef<SliderRef>();
    const { container, unmount } = render(
      <Slider ref={ref} />
    );
    fireEvent.mouseDown(container.querySelector('.handle'), { button: 0 });
    expect(ref.current.moving).toBe(true);

    fireEvent.mouseUp(document.body);
    expect(ref.current.moving).toBe(false);

    fireEvent.mouseDown(container.querySelector('.handle'), { button: 1 });
    expect(ref.current.moving).toBe(false);

    fireEvent.mouseUp(document.body);
    expect(ref.current.moving).toBe(false);

    unmount();
  });

  it('should trigger onMove event on mouse move', async () => {
    const ref = createRef<SliderRef>();
    const onMove = vi.fn();
    const { container, unmount } = render(
      <Slider ref={ref} onMove={onMove} min={0} max={100} step={1} />
    );

    await act(async () => {
      Object.defineProperty(
        ref.current.slideRef.current,
        'offsetWidth',
        { value: 234 }
      );

      return Promise.resolve();
    });
    fireEvent.mouseDown(container.querySelector('.handle'), { button: 0 });

    const moveEvent = createEvent
      .mouseMove(container.querySelector('.handle'));
    Object.defineProperty(moveEvent, 'pageX', { value: 25 });
    fireEvent(container.querySelector('.handle'), moveEvent);

    expect(onMove).toHaveBeenCalledWith(11);

    unmount();
  });

  it('shouldn\'t trigger onMove event on mouse move when slider is ' +
    'disabled', () => {
    const ref = createRef<SliderRef>();
    const onMove = vi.fn();
    const { container, unmount } = render(
      <Slider
        ref={ref}
        globalEventsTarget={document}
        disabled={true}
        onMove={onMove}
      />
    );
    fireEvent.mouseDown(container.querySelector('.handle'), { button: 0 });
    expect(ref.current.moving).toBe(false);

    const moveEvent = createEvent
      .mouseMove(container.querySelector('.handle'));
    Object.defineProperty(moveEvent, 'pageX', { value: 25 });
    fireEvent(container.querySelector('.handle'), moveEvent);

    expect(onMove).not.toHaveBeenCalled();
    unmount();
  });

  it('should allow to move slider using arrow keys', () => {
    const ref = createRef<SliderRef>();
    const { container, unmount } = render(
      <Slider
        ref={ref}
        min={0}
        max={100}
        step={10}
        value={0}
      />
    );

    fireEvent
      .keyDown(container.querySelector('.handle'), { key: 'ArrowRight' });
    expect(ref.current.value).toBe(10);

    fireEvent
      .keyDown(container.querySelector('.handle'), { key: 'ArrowRight' });
    expect(ref.current.value).toBe(20);

    fireEvent
      .keyDown(container.querySelector('.handle'), { key: 'ArrowLeft' });
    expect(ref.current.value).toBe(10);

    fireEvent
      .keyDown(container.querySelector('.handle'), { key: 'ArrowDown' });
    expect(ref.current.value).toBe(100);

    fireEvent
      .keyDown(container.querySelector('.handle'), { key: 'ArrowUp' });
    expect(ref.current.value).toBe(0);

    fireEvent
      .keyDown(container.querySelector('.handle'), { key: 'PageUp' });
    expect(ref.current.value).toBe(0);

    unmount();
  });
});
