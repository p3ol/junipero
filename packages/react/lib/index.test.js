import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
