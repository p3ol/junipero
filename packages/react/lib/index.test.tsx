import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
