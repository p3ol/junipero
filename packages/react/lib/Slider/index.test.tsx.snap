// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Slider /> should render 1`] = `
<div>
  <div
    class="junipero slider"
  >
    <div
      class="slide"
    >
      <div
        class="fill"
        style="width: 0%;"
      />
    </div>
    <div
      class="handle"
      style="transform: translate3d(0px, 0, 0);"
      tabindex="0"
    />
  </div>
</div>
`;
